### DISCUSSION TITLE
Handling Route Execution Order and Conditional Middleware in Mojo.js

### DISCUSSION THREAD

**Discussion by xdaruis** [mojolicious/mojo.js#162] [Q&A]
Hey, I want to implement the following feature in the backend using `mojojs`. I want the routes to be executed from top to bottom and have a helper function to determine logged-in routes.

I have a problem with `app.under` and how I've used it, and now the 2nd route gets executed before the first one.

```js
const app = mojo();

const loggedIn = app.under('', async (ctx) => {
  const session = await ctx.session();
  if (!session || !session.role) {
    await ctx.render({ status: 401, json: { error: 'Forbidden' } });
    return false;
  }
});

app.get('/api/controller/stats').to('controller#stats'); // Doesn't work
loggedIn.get('/api/controller/:id').to('controller#get'); // This gets executed before the above route
```

I managed to use the router and requires to fix the problem, but now I have another issue in which the request is still trying to get picked up by all the next routes.

```js
const app = mojo();
const { router } = app;

router.addCondition('loggedIn', async (ctx) => {
  const session = await ctx.session();
  if (!session || !session.role) {
    await ctx.render({ status: 401, json: { error: 'Forbidden' } });
    return false;
  }
  return true;
});

// Gets executed
router.get('/api/controller/action').requires('loggedIn').to('controller#action');

// I don't want this to get executed anymore since the request is already handled, but it does
router.get('/*', async (ctx) => {
  console.log('Why am i seeing this? :((');
  return ctx.render({ status: 500, json: { error: 'I dont wanna see this' } });
});
```

**Comment by kraih** [comment]
Only one route will match the request. Unlike other frameworks, mojo.js does not execute all matching routes, only the first one. And code is only executed when the full path matches, including nested routes.

**Reply by kraih** [reply]
Or to be more precise, this is not a Sinatra-like framework where all code runs. We build a tree structure out of the declared routes, and only the first fully matching branch gets its code executed.

**Reply by xdaruis** [reply]
Sorry if I wasn't so clear; my problem isn't about allowing multiple routes to get triggered. I just want a way to safely execute routes from top to bottom, and the first matching route should be executed.

In the first code, somehow the `loggedIn` route gets executed first, and then the `app.get(../stats)` will never be executed because of the `:id` in the loggedIn route.

The second code using the router gets the job done, but now I would get the `/api/controller/action` response and then also a warning in the console, and that `console.log()` would be printed in the console.

Here's the error I'm getting in the 2nd code snippet:
```
Error [ERR_HTTP_HEADERS_SENT]: Cannot write headers after they are sent to the client
    at ServerResponse.writeHead (node:_http_server:344:11)
```

Not sure if this is important but I'm running the server in development mode:

```
nodemon index.js server -l http://*:8086
```

**Reply by kraih** [reply]
Now i get what you mean. Conditions are not meant to have side effects such as rendering a response themselves. What happens when they do is undefined behavior.

**Reply by kraih** [reply]
For your `/api/controller/:id` case, you could just make the placeholder more restrictive so it only matches numbers. https://mojojs.org/docs/Routing.md#restrictive-placeholders

**Reply by kraih** [reply]
Or, of course, just change the order so the precedence of the routes changes.

**Reply by xdaruis** [reply]
"Now i get what you mean. Conditions are not meant to have side effects such as rendering a response themselves. What happens when they do is undefined behavior." -> I see, thanks for the good info on this.

Apparently, I've figured out the behavior with `app.under('/')`. The routes get grouped in regex matches, so in my first example, my welcome would never get used since `loggedIn` is a priority group as seen in the picture.

```js
const loggedIn = app.under('/', async (ctx) => {
  const session = await ctx.session();
  if (!session || !session.role) {
    await ctx.render({ status: 401, json: { error: 'Forbidden' } });
    return false;
  }
});

app.get('/api/controller/welcome').to('example#welcome');
loggedIn.get('/api/controller/:uuid').to('example#get');
```

<img width="974" alt="image" src="https://github.com/user-attachments/assets/23721573-c96d-4a9a-becd-62178d81e471">

Here's a solution I've figured out:

```js
const myRouter = app.under('', async () => true);

const loggedIn = app.under('/', async (ctx) => {
  const session = await ctx.session();
  if (!session || !session.role) {
    await ctx.render({ status: 401, json: { error: 'Forbidden' } });
    return false;
  }
});

myRouter.get('/api/controller/welcome').to('example#welcome');
loggedIn.get('/api/controller/:uuid').to('example#get');
```

<img width="987" alt="image" src="https://github.com/user-attachments/assets/a0860b3e-cd08-4903-abde-b8244070330e">


Thanks for all the help so far, and if you have a more elegant solution than this, please let me know.

**Reply by kraih** [reply]
There is no need for an empty handler, nested routes are a fundamental concept of our router: https://mojojs.org/docs/Routing.md#nested-routes

**Reply by xdaruis** [reply]
I'm still a bit confused about how to avoid empty handlers while still using the powers of the `loggedIn` and `isAdmin` middleware and avoid ordering conflicts as presented above.

Here are some routes which I would like to point to the same controller (some of them can have the `/api/admin` prefix), but for the logged in routes I'm creating an empty route with a callback to check if the user is logged in:

```js
const loggedIn = async (ctx) => {
  const session = await ctx.session();
  if (!session || !session.role) {
    await ctx.render({ status: 401, json: { error: 'Forbidden' } });
    return false;
  }
};

const isAdmin = app.under('/api/admin', async (ctx) => {
  const session = await ctx.session();
  if (!session || session.role !== 'ADMIN') {
    await ctx.render({ status: 403, json: { error: 'Unauthorized' } });
    return false;
  }
});

const controller = app.any('/api/controller').to('controller');
const admin_controller = isAdmin.any('/controller').to('controller');

controller.get('/welcome').to('#welcome');
controller.get('/:uuid').under('', loggedIn).to('#get');
controller.post('/:uuid').under('', loggedIn).to('#create');
controller.delete('/:uuid/delete').under('', loggedIn).to('#delete');

admin_controller.get('/list').to('#list');
```

Here's a picture of the new routes but it creates the empty routes in the file: `Pattern +/ Method *`. Is this implementation good enough in order to keep my `loggedIn` middleware?



<img width="708" alt="image" src="https://github.com/user-attachments/assets/28f760e9-6c9b-4e8e-bb54-4160faadea31">
